### ISSUE TITLE
split train/test by users

### ISSUE THREAD

**Issue by ziliangok** [tensorflow/recommenders#123]
I have run this examples https://tensorflow.google.cn/recommenders/examples/multitask?hl=en
at the data splitting part, it using a random splitting method, does tfrs provide a method to partition dataset user-by-user. For example, random select 20% data for each user as the test set.

**Comment by maciejkula** [comment]
This should be fairly straightforward using standard `tf.data.Dataset` manipulations.

For example, to deterministically split by user id you might do the following using `mod` as your hashing function:

```python
train = data.filter(lambda x: tf.math.floormod(x["user_id"], 100) < 80)
test = data.filter(lambda x: tf.math.floormod(x["user_id"], 100) >= 80)
```

If your user ids are strings and not numbers, [this function](https://www.tensorflow.org/api_docs/python/tf/strings/to_hash_bucket) might help you do the same.

**Comment by ziliangok** [comment]
> This should be fairly straightforward using standard `tf.data.Dataset` manipulations.
> 
> For example, to deterministically split by user id you might do the following using `mod` as your hashing function:
> 
> ```python
> train = data.filter(lambda x: tf.math.floormod(x["user_id"], 100) < 80)
> test = data.filter(lambda x: tf.math.floormod(x["user_id"], 100) >= 80)
> ```
> 
> If your user ids are strings and not numbers, [this function](https://www.tensorflow.org/api_docs/python/tf/strings/to_hash_bucket) might help you do the same.

Thank you for your answer, but it looks like not what i expected. I mean for each user, select 80% as the training set. For example, the user 1 have 100 records, so we pick 80, and the user 2 have 200 records, and we pick 160. 

In your code, it looks like the user whose id between x00-x79 will be the train set, and the others will be the test set.

**Comment by maciejkula** [comment]
What you suggest is likely very close to what the existing tutorials are doing: they take 80% of all interactions for the train set, and 20% for the test set. On average, this means that 80% of any user's interactions are in the training set and 20% in the testing set.

Doing exactly the split you suggest will be reasonably tricky and I'm not sure will bring tangible benefits.

**Comment by ziliangok** [comment]
I found a method using pandas groupby+sampe but require convert Tensor to DataFrame and then convert DataFrame to Datasets. Or using Dataset.fiter+shuffle with loopping each user, these could be very tricky.

My thought is using this partition could get more 'balanced' training and testing datasets.

**Comment by maciejkula** [comment]
If you find a solution, please post it here. My advice would be not to spend too much effort on it: it's unlikely to be important to the quality of your model.

What I would suggest instead is time-based test/train splits: all interactions before a certain time to go the train set, and all later interactions go to the test set. This will be closer to a production-like setting where you use data up to today to make recommendations tomorrow.

**Comment by ziliangok** [comment]
```python
df = tfds.as_dataframe(ratings)
test_index = df.groupby('user_id').sample(frac=0.2).index
train_index = set(df.index) - set(test_index)
train_df = df.loc[train_index]
test_df = df.loc[test_index]
# convert DataFrame to Dataset
'''
you are right, time-based splitting could be more meaningful, and this is not very important.
thank you a lot.
